tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc

```diff
@@ -0,0 +1,266 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "mlir-hlo/Analysis/userange_analysis.h"
+#include "mlir-hlo/Transforms/PassDetail.h"
+#include "mlir-hlo/Transforms/passes.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/Interfaces/LoopLikeInterface.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/BufferUtils.h"
+
+namespace mlir {
+
+namespace {
+
+/// Reuses already allocated buffer to save allocation operations.
+class BufferReuse : BufferPlacementTransformationBase {
+ public:
+  BufferReuse(Operation* op)
+      : BufferPlacementTransformationBase(op),
+        dominators(op),
+        postDominators(op),
+        userange(op, allocs, aliases) {}
+
+  /// Reuses already allocated buffers to save allocation operations.
+  void reuse(Operation* operation) {
+    // Create a list of values that can potentially be replaced for each value
+    // in the useRangeMap. The potentialReuseMap maps each value to the
+    // respective list.
+    llvm::MapVector<Value, SmallVector<Value, 4>> potentialReuseMap;
+    for (BufferPlacementAllocs::AllocEntry entry : allocs) {
+      Value itemA = std::get<0>(entry);
+      SmallVector<Value, 4> potReuseVector;
+      for (BufferPlacementAllocs::AllocEntry entry : allocs) {
+        Value itemB = std::get<0>(entry);
+        // Do not compare an item to itself and make sure that the value of item
+        // B is not a BlockArgument. BlockArguments cannot be reused. Also
+        // perform a type check.
+        if (itemA == itemB || !checkTypeCompatibility(itemA, itemB)) continue;
+
+        // Check if itemA can replace itemB.
+        if (!userange.rangesInterfere(itemA, itemB)) continue;
+
+        // Get the defining block of itemA.
+        Block* defOpBlock = itemA.isa<BlockArgument>()
+                                ? itemA.getParentBlock()
+                                : itemA.getDefiningOp()->getBlock();
+
+        // The defining block of itemA has to dominate all uses of itemB.
+        if (!dominatesAllUses(defOpBlock, itemB)) continue;
+
+        // Insert itemB into the right place of the potReuseVector. The order of
```

Comment:
It would be interesting to explore other orders here. This is essentially greedy but one could try prioritize reuses for certain cases. Like if one has a copy, then source and destination should preferably be reused.